### DISCUSSION TITLE
Post запрос к виртуальной АТС

### DISCUSSION THREAD

**Discussion by papaoleg** [totumonline/totum-ru-issues-and-discussions#515] [Как это сделать?] [answered]
Добрый день.
Очень прошу помощи, мозг уже взрывается, не получается.
В общем, сейчас хочу реализовать функцию вызова звонка через нажатие кнопки (оператор [МСН Телеком](https://www.helpmcn.ru/ru-RU/support/solutions/articles/43000718133-%D0%B2%D1%8B%D0%B7%D0%BE%D0%B2-%D0%B8%D1%81%D1%85%D0%BE%D0%B4%D1%8F%D1%89%D0%B5%D0%B3%D0%BE-%D0%B7%D0%B2%D0%BE%D0%BD%D0%BA%D0%B0-%D1%81-%D0%B2%D0%BD%D1%83%D1%82%D1%80%D0%B5%D0%BD%D0%BD%D0%B5%D0%B3%D0%BE-%D0%BD%D0%BE%D0%BC%D0%B5%D1%80%D0%B0-%D0%B2%D0%B0%D1%82%D1%81-%D1%87%D0%B5%D1%80%D0%B5%D0%B7-api)).
Отправляю запрос через
`getFromScript(uri: "https://vpbx.mcn.ru/api/protected/calls/vpbx/outbound"; posts: $post; headers: $head; method: "POST")`
Тело запроса, который прекрасно срабатывает в постмане и reqbin выглядит так:

> POST /api/protected/calls/vpbx/outbound HTTP/1.1
Authorization: Bearer *************************
Host: vpbx.mcn.ru
Accept: application/json
Content-Type: application/json
Content-Length: 91

> {
  "timeoutFrom": 0,
  "vpbxId": 1111,
  "fromNumber": "100",
  "toNumber":"79271234567"
}

В случае правильного исполнения в ответ приходит айдишник совершенного вызова, например:
`"originateId": "78c5a28e-37f0-4318-a95a-dcab72e9c741"`

Пробовал разные варианты:
```
getFromScript(uri: "https://vpbx.mcn.ru/api/protected/calls/vpbx/outbound"; posts: $post; headers: $head; method: "POST"; header: true)
	head: listCreate(item: "Authorization: Bearer *********"; item: "Host: vpbx.mcn.ru"; item: "Content-Length: 75")
	post: jsonCreate(field: "timeoutFrom" = 0; field: "vpbxId" = #nomer_vats; field: "fromNumber" = "100"; field: "toNumber" = #phone)
``` 
и вот так:
```
getFromScript(uri: "https://vpbx.mcn.ru/api/protected/calls/vpbx/outbound"; post: "timeoutFrom" = 0; post: "vpbxId" = 1111; post: "fromNumber" = "100"; post: "toNumber" = "79271234567"; headers: $head; method: "POST")
	head: listCreate(item: "Authorization: Bearer *********"; item: "Host: vpbx.mcn.ru"; item: "Content-Type: application/json"; item: "Content-Length: 61")
``` 
Не срабатывает.
В связи с чем, несколько вопросов:
1. Как узнать какой ответ приходит от сервера? И как куда-нибудь воткнуть потом полученный айдишник звонка?
2. Сервер очень чувствителен к Content-Length - его неправильное значение всё убивает. Как правильно его вычислять для подстановки в getFromScript?
3. Ну и главное - почему не срабатывает? Может я где-то что-то упустил или неправильно заполнил (глаз уже замылился)? Еще в запросе поля timeoutFrom и vpbxId должны быть числами, а остальные строчками - всё ли верно указано в функции?

Заранее благодарю за помощь.

**Comment by erpic-ru** [comment]
Узнать легко:
1. Тестовые запросы через postman
2. Создаете доп. поле "Данные" и сетите в него результат

```
set(table: $#ntn; field: 'h_test' = $json; )
	json: jsonExtract(text: $getfromscript)
```

**Reply by erpic-ru** [reply]
Первый вариант более правильный, но в нем вы не передаете заголовок: "Content-Type: application/json"; это тоже может вызывать ошибку

**Reply by erpic-ru** [reply]
А еще есть фокус с числами, тк в тотуме все строки. Если это критично, то заверните их в спец функцию
https://ru.docs.totum.online/functions#fn-numTransform

**Reply by papaoleg** [reply]
В постмане всё работает. 
Почему-то не сетится:
`Ошибка разбора JSON: Syntax error.
Функция jsonExtract; Линия кода json.; Set field 'h_rez'=$json; Функция set; action поле zvonok_test_1 таблицы telefonnye_zvonki id 123`

`set(table: $#ntn; field: 'h_rez' = $json)
json: jsonExtract(text: $get)`

**Reply by erpic-ru** [reply]
ну попробуйте без json, сетить результат gfs

**Reply by erpic-ru** [reply]
+ header: false можно сделать

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Так, всегда с АТС какая-то хрень. Не пойму, то ли от инструментов на которых они там програмят, то-ли от того, что они нанимают каких-то последователей жОсткОй типизации.

Первое. делаем нам сначала вывод ответа сервера.

```
= : linkToDataText(title: "*"; text: $gfs)

gfs: getFromScript(uri: "https://vpbx.mcn.ru/api/protected/calls/vpbx/outbound"; posts: $post; headers: $head; header: true)
	head: listCreate(item: "Authorization: Bearer *********"; item: "Host: vpbx.mcn.ru"; item: "Content-Length: 75")
	post: jsonCreate(field: "timeoutFrom" = 0; field: "vpbxId" = 444555; field: "fromNumber" = "100"; field: "toNumber" = "79255006080")
```
+ method для post можно не передавать (он при использовании posts и так будет post)
+ header: true — это нужное для отладки, тк показывает заголовки
+ linkToDataText тк ответ приходит текстом

Запускаем:

```
HTTP/1.1 401 Unauthorized
date: Fri, 02 Feb 2024 11:32:26 GMT
content-type: text/html
content-length: 574
strict-transport-security: max-age=15724800; includeSubDomains
```

Понятно, что авторизация накернилась тк у меня там `****`. Но теперь мы видим ошибки от стороннего сервера.

Докручиваем код:

```
= : linkToDataText(title: "*"; text: $gfs)

gfs: getFromScript(uri: "https://vpbx.mcn.ru/api/protected/calls/vpbx/outbound"; posts: $post; headers: $head; header: true)
	head: listCreate(item: "Authorization: Bearer *********"; item: "Host: vpbx.mcn.ru"; item: str`"Content-Length:" ++ $length`; item: "Authorization: Bearer *********"; item: "Content-Type: application/json")
	length: strLength(str: $post)
		~post: jsonCreate(field: "timeoutFrom" = 0; field: "vpbxId" = 444555; field: "fromNumber" = "100"; field: "toNumber" = "79255006080")
```

+ "Content-Type: application/json" — передаем в заголовки, тк по умолчанию там application/x-www-form-urlencoded

+ str`"Content-Length:" ++ $length` — считаем длину передаваемой строки (если я правильно понимаю о чем идет речь) ф. strLength

+ ~post: — ~ фиксируем первое вычисление, тк у нас два запроса к этой строке и никакой динамики между этими запросами нет.

+ field: "timeoutFrom" = 0;  — когда мы вот так 0 указываем, он будет в json как число (без кавычек)

+ field: "timeoutFrom" = #timeout и в поле timeout лежит 0 то в итоговом json будет "0" тк в тотуме все хранится в виде строки. Некоторые API не пропускают это (ж.типизация). Тогда преобразуем его через numTransform
    ```
    // example
    
    ~post: jsonCreate(field: "timeoutFrom" = $timeout)
			timeout: numTransform(data: #timeout)
    ```
    numTransform может и массивы обрабатывать — посмотрите доки
    
Дальше я не могу проверить, тк авторизации нет.

**Reply by ttm-alexeypolunin** [reply]
Еще хорошая идея:

+ сделать на стороне тотума флаг, который выводит не ответ, а отправляемый запрос.
+ и еще надо header: true доапгрейдить, что бы автоматически был массив headers/body с раскладыванием по ключам (но это не точно)

**Comment by papaoleg** [comment]
Невероятно! Оно заработало! После того, как появилась возможность видеть ошибку, стало понятно, что проблема в нуле и vpbxId - при передаче 'timeoutFrom': 0 их сервер всё равно понимал 'timeoutFrom': "0", поэтому пришлось оборачивать в numTransform даже ноль и 1111.
Огромное спасибо за помощь!

**Reply by ttm-alexeypolunin** [reply]
трайпл килл :)
